src/main/groovy/com/github/winplay02/gitcraft/MinecraftVersionGraph.java

```diff
@@ -105,39 +103,118 @@ private void testGraphConnectivity() {
+
+				if (nextBranchLength > branchLength) {
+					branchLength = nextBranchLength;
+				}
 			}
-			MiscHelper.panic(amountRootNodes < 1 ? "There is no root node. This either means, that the version graph is empty, or that it contains a cycle." : "There are multiple root nodes. A connected git history would not be guaranteed");
 		}
+
+		return branchLength + 1;
 	}
 
-	protected static final Pattern LINEAR_SNAPSHOT_REGEX = Pattern.compile("(^\\d\\dw\\d\\d[a-z]$)|(^\\d.\\d+(.\\d+)?(-(pre|rc)\\d+|_[a-z_\\-]+snapshot-\\d+| Pre-Release \\d+)?$)");
+	private void markMainBranch(OrderedVersion mcVersion) {
+		// after the branch points have been identified,  search for the longest path
+		// from each of the roots to a tip, and mark that path as a main branch - that
+		// is to say, remove any branch points that lay on these paths from the map,
+		// as it should only contain side branches
+
+		NavigableSet<OrderedVersion> nextBranches = this.getFollowingNodes(mcVersion);
+
+		OrderedVersion mainBranch = null;
+		int mainBranchLength = -1;
+
+		for (OrderedVersion branch : nextBranches) {
+			int branchLength = this.branchPoints.getOrDefault(branch, 0);
+
+			if (branchLength > mainBranchLength) {
+				mainBranch = branch;
+				mainBranchLength = branchLength;
+			}
+		}
+
+		if (mainBranch != null) {
+			if (mainBranchLength > 0) {
+				this.branchPoints.remove(mainBranch);
+			}
 
-	public static boolean isVersionNonLinearSnapshot(OrderedVersion mcVersion) {
-		// remove all pending "snapshots" from mainline and mark as non-linear
-		return mcVersion.isPending() || mcVersion.isSnapshotOrPending() && (Objects.equals(mcVersion.launcherFriendlyVersionName(), "15w14a") || !(LINEAR_SNAPSHOT_REGEX.matcher(mcVersion.launcherFriendlyVersionName()).matches())); // mark 15w14a explicit as april fools snapshot, since this case should not be covered by the regex
+			this.markMainBranch(mainBranch);
+		}
 	}
 
 	public HashSet<String> repoTags = new HashSet<>();
+	/** root nodes of the graph, mapped to the path lengths to the tips of those main branches */
+	public HashMap<OrderedVersion, Integer> roots = new HashMap<>();;
 	public HashMap<OrderedVersion, TreeSet<OrderedVersion>> edgesBack = new HashMap<>();
 	public HashMap<OrderedVersion, TreeSet<OrderedVersion>> edgesFw = new HashMap<>();
+	/** nodes that mark new side branches, mapped to the path lengths to the tip sof those side branches */
+	public HashMap<OrderedVersion, Integer> branchPoints = new HashMap<>();
 
-	public static MinecraftVersionGraph createFromMetadata(ManifestSource manifestSource, ManifestProvider<?, ?> provider) throws IOException {
+	public static MinecraftVersionGraph createFromMetadata(MetadataProvider provider) throws IOException {
 		MinecraftVersionGraph graph = new MinecraftVersionGraph();
-		if (manifestSource != ManifestSource.MOJANG_MINECRAFT_LAUNCHER) {
-			graph.repoTags.add(String.format("manifest_%s", manifestSource.toString()));
-		}
```

Comment:
This was added for compatibility reasons, as old repos would need to either be regenerated or renamed